### PR TITLE
Check for selection handler from field configuration

### PR DESCRIPTION
**Pull request by damiankloip**
…a query in OgSelection

Me and @pfresson saw this last week. It's pretty blocking, as it happens when creating any new nodes OR users. And that is just what I have tried it on. I think it will affect anything really.

I think part of the problem is that our selection handler is generic, so it doesn't apply to any particular type and it's in the default group, so maybe it gets chosen because of that, I think we might want to change the group too. This manifests itself when validators run, and selection handlers for the fields try to get referenceable entities to validate the entity before save.

This currently checks for a field first, as things break hard when that is not available. So at least we have something more robust.

I am changing the group too. That is actually the only fix that is needed to stop the fatals. Seems quite fragile though - if we were just to rely on the selection handler selection alone.

### CODE REVIEW

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 1]
```diff
@@ -25,7 +25,7 @@
  * @EntityReferenceSelection(
  *   id = "og:default",
  *   label = @Translation("OG selection"),
- *   group = "default",
+ *   group = "og",
```
Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:28]: why this change?

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 2]
```diff
@@ -60,7 +64,14 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
     // the default selection handler of the entity, which the field reference
     // to, and add another logic to the query object i.e. check if the entities
     // bundle defined as group.
-    $query = $this->getSelectionHandler()->buildEntityQuery();
+    $selection_handler = $this->getSelectionHandler();
+
+    if (empty($selection_handler)) {
+      return $query = parent::buildEntityQuery($match, $match_operator);
```
Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:70]: Maybe throw an exception instead? Otherwise we silently fail here.

Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:70]: Ha, that's exactly what we need to avoid here :)

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 1 above)
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:28]: Because default is well..the default so when the selection handler tried to select the 'best' handler, it comes back with our OG one regardless of entity type. This is why creating a user entity is currently fataling too. See `SelectionPluginManager::getPluginId` and `SelectionPluginManager::getSelectionGroups`

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 2 above)
Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:70]: not sure we want to avoid it - can you explain why?

The way I see it - if the `$selection_handler` is missing it means something in the configuration is missing, no?

Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:70]: That will still leave us with a fatal as well.

Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:70]: > That will still leave us with a fatal as well.

Indeed, which If I understand the scenario correctly, I think is the right outcome. If the field you expected to have is badly configured - you'd want to know about it loudly. Or am I missing something? :)

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 1 above)
Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:28]: > it comes back with our OG one regardless of entity type

This is the expected behavior - our selection handler is a wrapper around the original - entity specific - selection handler.

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 2 above)
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:70]: I think so, as the field could not always be configured, like when it is called for validation constraints. See the `ValidReferenceContraintValidator`. If there is not a field, there is no way the `OgSelection` handler can get the base handler. Other handler don't work this way or rely on this value, so it's not a problem.

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 1 above)
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:28]: OK, so if we are doing that, then we DEFINITELY need the below change to not throw anthing :) Where would that behaviour of always selecting the of handler get used?

Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:28]: > OK, so if we are doing that, then we DEFINITELY need the below change to not throw anthing

Can you provide ways to reproduce the problem - I'd like to better understand the scenario, I have a feeling I'm missing your point ;)

> Where would that behaviour of always selecting the of handler get used?

https://github.com/amitaibu/og/blob/8.x-1.x/src/Plugin/EntityReferenceSelection/OgSelection.php#L63

We get the "original" (entity specific) query, and then we add conditions to it.

Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:28]: Yes, but if we don't have the field, we can't get a selection handler based on field. For me this is happening when trying to create any node or user. Probably any other entity possibly that will use the valid reference constraint.

Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:28]: I don't mean the funcitonality of the handler, that I get ok. I mean where would it rely on being in the default group for handler selection?

Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:28]: > I mean where would it rely on being in the default group for handler selection?

Oh, that is not needed.

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 2 above)
Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:70]: Ok, thanks - I'll check later what you mean with the constraints

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 1 above)
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:28]: Cool, so I think using our own OG group in the annotation should stay in that case.

**Comment by pfrenssen** [comment]
Great, thank for working on this, this blocked me last week. I'll hopefully have some time tomorrow to test it out.
